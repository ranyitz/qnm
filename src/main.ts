import merge from 'lodash/merge';
import Workspace from './workspace/workspace';
import { CliOptions } from './cli';

import matchAction from './actions/match';
import getAction from './actions/get';
import listAction from './actions/list';
import fuzzySearchAction from './actions/fuzzy-search';

type Options = Pick<CliOptions, 'deps' | 'match'>;

const defaultOptions = {
  noColor: true,
};

export function match(matchPattern: string, cwd?: string, options?: Options) {
  const workspace = Workspace.loadSync(cwd);

  return matchAction(workspace, matchPattern, merge(defaultOptions, options));
}

export function get(name: string, cwd?: string, options?: Options) {
  const workspace = Workspace.loadSync(cwd);

  return getAction(workspace, name, merge(defaultOptions, options));
}

export function list(cwd?: string, options?: Options) {
  const workspace = Workspace.loadSync(cwd);

  return listAction(workspace, merge(defaultOptions, options));
}

export function search(cwd?: string, options: Options = {}) {
  const workspace = Workspace.loadSync(cwd);

  return fuzzySearchAction(workspace, merge(defaultOptions, options));
}
