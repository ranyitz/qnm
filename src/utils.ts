export function isTruthy<T>(x: T | undefined | null): x is T {
  return x !== undefined && x !== null;
}
