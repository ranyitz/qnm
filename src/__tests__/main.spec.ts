import * as qnm from '../main';
import { resolveFixture } from './utils';

describe('Node API', () => {
  describe('qnm <module>]', () => {
    it('should show the version and dependents info on a single module when called with a string', () => {
      const cwd = resolveFixture('single-module');
      const output = qnm.match('test', cwd);

      expect(output).toMatchSnapshot();
    });

    it('should show get matches when using the match command', () => {
      const cwd = resolveFixture('single-module');
      const output = qnm.match('te', cwd);

      expect(output).toMatchSnapshot();
    });

    it('should add dependents information on yarn installed package', () => {
      const cwd = resolveFixture('yarn-install');
      const output = qnm.match('import-from', cwd);

      expect(output).toMatchSnapshot();
    });

    it('should show an indication in case there is a symlink', () => {
      const cwd = resolveFixture('symlink');
      const output = qnm.match('test', cwd);

      expect(output).toMatchSnapshot();
    });

    it('should work in monorepo and print subpackages modules', () => {
      const cwd = resolveFixture('monorepo');
      const output = qnm.match('package-foo', cwd);

      expect(output).toMatchSnapshot();
    });

    it('should work in monorepo with yarn workspaces', () => {
      const cwd = resolveFixture('monorepo-with-workspaces');
      const output = qnm.match('package-foo', cwd);

      expect(output).toMatchSnapshot();
    });

    it('should provide suggestion for scoped package with the same name', () => {
      const cwd = resolveFixture('scoped-package');

      try {
        qnm.match('test', cwd);
      } catch (error) {
        expect(error.message).toMatch('Did you mean "@scope/test"');
      }
    });
  });

  describe('qnm list', () => {
    it('should show all modules in node_modules directory', () => {
      const cwd = resolveFixture('mix-modules');
      const output = qnm.list(cwd);

      expect(output).toMatchSnapshot();
    });

    it('should show modules mentioned in package.json', () => {
      const cwd = resolveFixture('indirect-dependencies');
      const output = qnm.list(cwd, { deps: true });

      expect(output).toMatchSnapshot();
    });

    it('should list dependencies in a yarn installed package and show "why" information', () => {
      const cwd = resolveFixture('yarn-install');
      const output = qnm.list(cwd);

      expect(output).toMatchSnapshot();
    });

    it('should list a monorepo', () => {
      const cwd = resolveFixture('monorepo');
      const output = qnm.list(cwd);

      expect(output).toMatchSnapshot();
    });
  });

  describe('qnm match', () => {
    it('should match in monorepo and print subpackages modules', () => {
      const cwd = resolveFixture('monorepo');
      const output = qnm.match('packa', cwd);

      expect(output).toMatchSnapshot();
    });
  });
});
