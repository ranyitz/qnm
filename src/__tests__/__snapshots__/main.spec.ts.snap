// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node API qnm <module>] should add dependents information on yarn installed package 1`] = `
"[4m[35mimport-from[39m[24m
â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/yarn-install/node_modules/import-from/package.json3.0.0]8;; [33m(import-cwd)[39m
"
`;

exports[`Node API qnm <module>] should show an indication in case there is a symlink 1`] = `
"[4m[35mtest[39m[24m
â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/symlink/node_modules/test/package.json1.0.0]8;;[35m -> ../../symlink-origin/node_modules/test[39m [33m(devDependencies, npm install test)[39m
"
`;

exports[`Node API qnm <module>] should show get matches when using the match command 1`] = `
"[4m[35mte[39mst[24m
â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/single-module/node_modules/test/package.json1.0.0]8;; [33m(devDependencies, npm install test)[39m
"
`;

exports[`Node API qnm <module>] should show the version and dependents info on a single module when called with a string 1`] = `
"[4m[35mtest[39m[24m
â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/single-module/node_modules/test/package.json1.0.0]8;; [33m(devDependencies, npm install test)[39m
"
`;

exports[`Node API qnm <module>] should work in monorepo and print subpackages modules 1`] = `
"[4m[35mpackage-foo[39m[24m
â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/monorepo/node_modules/package-foo/package.json1.0.0]8;;
"
`;

exports[`Node API qnm <module>] should work in monorepo with yarn workspaces 1`] = `
"[4m[35mpackage-foo[39m[24m
â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/monorepo-with-workspaces/node_modules/package-foo/package.json1.0.0]8;;
"
`;

exports[`Node API qnm list should list a monorepo 1`] = `""`;

exports[`Node API qnm list should list dependencies in a yarn installed package and show "why" information 1`] = `""`;

exports[`Node API qnm list should show all modules in node_modules directory 1`] = `
"[4m@scope/test[24m
â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/mix-modules/node_modules/@scope/test/package.json1.0.0]8;;

[4manother[24m
â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/mix-modules/node_modules/another/package.json1.0.0]8;;

[4mtest[24m
â”œâ”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/mix-modules/node_modules/test/package.json1.0.0]8;;
â””â”€â”¬ [90manother[39m
  â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/mix-modules/node_modules/another/node_modules/test/package.json1.0.0]8;;
"
`;

exports[`Node API qnm list should show modules mentioned in package.json 1`] = `
"[4mdependency1[24m
â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/indirect-dependencies/node_modules/dependency1/package.json1.0.0]8;;

[4mdependency2[24m
â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/indirect-dependencies/node_modules/dependency2/package.json1.0.0]8;;

[4mdevDependency1[24m
â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/indirect-dependencies/node_modules/devDependency1/package.json1.0.0]8;;

[4mdevDependency2[24m
â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/indirect-dependencies/node_modules/devDependency2/package.json1.0.0]8;;
"
`;

exports[`Node API qnm match should match in monorepo and print subpackages modules 1`] = `
"[4m[35mpacka[39mge-bar[24m
â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/monorepo/node_modules/package-bar/package.json1.0.0]8;;

[4m[35mpacka[39mge-baz[24m
â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/monorepo/node_modules/package-baz/package.json1.0.0]8;;

[4m[35mpacka[39mge-foo[24m
â””â”€â”€ ]8;;File:/Users/deansh/Projects/qnm/src/__tests__/fixtures/monorepo/node_modules/package-foo/package.json1.0.0]8;;
"
`;
